src/msha/auth/routes/auth-login-provider-callback.ts

```diff
@@ -0,0 +1,480 @@
+      });
+    }
+
+    if (verifiedEmail !== undefined) {
+      claims.push({
+        typ: "email_verified",
+        val: verifiedEmail,
+      });
+    }
+
+    if (name) {
+      claims.push({
+        typ: "name",
+        val: name,
+      });
+    }
+
+    if (picture) {
+      claims.push({
+        typ: "picture",
+        val: picture,
+      });
+    }
+
+    if (givenName) {
+      claims.push({
+        typ: "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname",
+        val: givenName,
+      });
+    }
+
+    if (familyName) {
+      claims.push({
+        typ: "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname",
+        val: familyName,
+      });
+    }
+
+    return {
+      identityProvider: "google",
+      userId,
+      userDetails,
+      claims,
+      userRoles: ["authenticated", "anonymous"],
+    };
+  } catch {
+    return null;
+  }
+};
+
+const getRoles = function (clientPrincipal: RolesSourceFunctionRequestBody, rolesSource: string) {
+  let cliApiUri = SWA_CLI_API_URI();
+  const { protocol, hostname, port } = parseUrl(cliApiUri);
+  const target = hostname === "localhost" ? `${protocol}//127.0.0.1:${port}` : cliApiUri;
+  const targetUrl = new URL(target!);
+
+  const data = JSON.stringify(clientPrincipal);
+
+  const options = {
+    host: targetUrl.hostname,
+    port: targetUrl.port,
+    path: rolesSource,
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      "Content-Length": Buffer.byteLength(data),
+    },
+  };
+
+  return new Promise((resolve, reject) => {
+    const protocolModule = targetUrl.protocol === "https:" ? https : http;
+
+    const req = protocolModule.request(options, (res) => {
+      res.setEncoding("utf8");
+      let responseBody = "";
+
+      res.on("data", (chunk) => {
+        responseBody += chunk;
+      });
+
+      res.on("end", () => {
+        try {
+          resolve(JSON.parse(responseBody));
+        } catch (err) {
+          reject(err);
+        }
+      });
+    });
+
+    req.on("error", (err) => {
+      reject(err);
+    });
+
+    req.write(data);
+    req.end();
+  });
+};
+
+const httpTrigger = async function (context: Context, request: http.IncomingMessage, customAuth?: SWAConfigFileAuth) {
+  const providerName = context.bindingData?.provider?.toLowerCase() || "";
+
+  if (providerName != "github" && providerName != "google") {
+    context.res = response({
+      context,
+      status: 404,
+      body: `Provider '${providerName}' not found`,
+    });
+    return;
+  }
+
+  const { cookie } = request.headers;
+
+  if (!cookie || !validateNonceCookie(cookie)) {
+    context.res = response({
+      context,
+      status: 401,
+    });
+    return;
+  }
+
+  const url = new URL(request.url!, `${SWA_CLI_APP_PROTOCOL}://${request?.headers?.host}`);
+
+  const codeValue = url.searchParams.get("code");
+  const stateValue = url.searchParams.get("state");
+
+  const nonce = decodeNonceCookie(cookie);
```

Comment:
the typical purpose of a nonce is to validate that it is what we get back at the end of the request as what we started with at the beginning of the loop, right? Does this achieve that? I may be confused but I'm not following what exact validation we are doing here

Reply:
This is verifying that the /callback call corresponds to the provider auth call that we initiated.

Before we redirect to https://github.com/login/oauth/authorize or https://accounts.google.com/o/oauth2/v2/auth, we generate a GUID, which we set in the "Nonce" cookie. We also hash this nonce and pass it as the "state" query string parameter in the provider auth URI that we're redirecting to.

Once the user has logged into the provider, the provider redirects them to our /callback API, passing "code" and "state" as query string parameters to the API. For "state" the provider is passing the same value that we originally passed when redirecting to the provider auth API.

So, if we receive a request to our /callback API, the value of the "state" query string parameter passed to the API should be the hash of the value in the "Nonce" cookie. We verify this by hashing the "Nonce" cookie value and comparing it to the "state" value. If the values don't match (or there is no "Nonce" cookie present), then we know this /callback call doesn't correspond to a provider auth redirect that we initiated and we fail the request with status 401.